src/lem/constrainer.rs

```diff
@@ -325,9 +828,184 @@ impl LEM {
                 _ => todo!(),
             }
         }
+
         if num_inputized_outputs != 3 {
             return Err(anyhow!("Couldn't inputize the right number of outputs"));
         }
+
+        if let Some(max_slots_indices) = max_slots_allowed {
+            if max_slots_indices.hash2_idx > hash_slots.hash2_data.max_slots {
+                bail!(
+                    "Too many slots allocated for Hash2/Unhash2: {}, {}",
+                    max_slots_indices.hash2_idx,
+                    hash_slots.hash2_data.max_slots
+                );
+            }
+            if max_slots_indices.hash3_idx > hash_slots.hash3_data.max_slots {
+                bail!(
+                    "Too many slots allocated for Hash3/Unhash3: {}, {}",
+                    max_slots_indices.hash2_idx,
+                    hash_slots.hash2_data.max_slots
+                );
+            }
+            if max_slots_indices.hash4_idx > hash_slots.hash4_data.max_slots {
+                bail!(
+                    "Too many slots allocated for Hash4/Unhash4: {}, {}",
+                    max_slots_indices.hash2_idx,
+                    hash_slots.hash2_data.max_slots
+                );
+            }
+        }
+
+        let alloc_dummy_ptr = alloc_manager.get_or_alloc_ptr(cs, &ZPtr::dummy())?;
+
+        /////////////////////////////////////////////////////////////////////////
+        ///////////////////////////// Hash2 /////////////////////////////////////
+        /////////////////////////////////////////////////////////////////////////
+
+        // Create hash constraints for each slot
+        {
+            let mut concrete_slots_hash2_len = 0;
+            let mut hash2_slots = HashMap::default();
+            while let Some((slot, alloc_car, alloc_cdr)) = hash_slots.hash2_alloc.pop() {
```

Comment:
If you're using a while loop for the purpose of iterating over a collection while holding an owned variant of the collection at each iteration, you're looking for `for element in collection.into_iter()` instead of `while let Some(element) = collection.pop()`.